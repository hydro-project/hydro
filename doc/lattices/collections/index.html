<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="Simple singleton or array collection with `cc_traits` implementations."><title>lattices::collections - Rust</title><script>if(window.location.protocol!=="file:")document.head.insertAdjacentHTML("beforeend","SourceSerif4-Regular-6b053e98.ttf.woff2,FiraSans-Italic-81dc35de.woff2,FiraSans-Regular-0fe48ade.woff2,FiraSans-MediumItalic-ccf7e434.woff2,FiraSans-Medium-e1aa3f0a.woff2,SourceCodePro-Regular-8badfe75.ttf.woff2,SourceCodePro-Semibold-aa29a496.ttf.woff2".split(",").map(f=>`<link rel="preload" as="font" type="font/woff2" crossorigin href="../../static.files/${f}">`).join(""))</script><link rel="stylesheet" href="../../static.files/normalize-9960930a.css"><link rel="stylesheet" href="../../static.files/rustdoc-20739d33.css"><meta name="rustdoc-vars" data-root-path="../../" data-static-root-path="../../static.files/" data-current-crate="lattices" data-themes="" data-resource-suffix="" data-rustdoc-version="1.88.0-nightly (10fa3c449 2025-04-26)" data-channel="nightly" data-search-js="search-f7877310.js" data-settings-js="settings-5514c975.js" ><script src="../../static.files/storage-4e99c027.js"></script><script defer src="../sidebar-items.js"></script><script defer src="../../static.files/main-7ef8a74a.js"></script><noscript><link rel="stylesheet" href="../../static.files/noscript-893ab5e7.css"></noscript><link rel="alternate icon" type="image/png" href="../../static.files/favicon-32x32-6580c154.png"><link rel="icon" type="image/svg+xml" href="../../static.files/favicon-044be391.svg"></head><body class="rustdoc mod"><!--[if lte IE 11]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><nav class="mobile-topbar"><button class="sidebar-menu-toggle" title="show sidebar"></button></nav><nav class="sidebar"><div class="sidebar-crate"><h2><a href="../../lattices/index.html">lattices</a><span class="version">0.6.1</span></h2></div><div class="sidebar-elems"><section id="rustdoc-toc"><h2 class="location"><a href="#">Module collections</a></h2><h3><a href="#structs">Module Items</a></h3><ul class="block"><li><a href="#structs" title="Structs">Structs</a></li><li><a href="#traits" title="Traits">Traits</a></li></ul></section><div id="rustdoc-modnav"><h2 class="in-crate"><a href="../index.html">In crate lattices</a></h2></div></div></nav><div class="sidebar-resizer"></div><main><div class="width-limiter"><rustdoc-search></rustdoc-search><section id="main-content" class="content"><div class="main-heading"><div class="rustdoc-breadcrumbs"><a href="../index.html">lattices</a></div><h1>Module <span>collections</span><button id="copy-path" title="Copy item path to clipboard">Copy item path</button></h1><rustdoc-toolbar></rustdoc-toolbar><span class="sub-heading"><a class="src" href="../../src/lattices/collections.rs.html#1-1203">Source</a> </span></div><details class="toggle top-doc" open><summary class="hideme"><span>Expand description</span></summary><div class="docblock"><p>Simple singleton or array collection with [<code>cc_traits</code>] implementations.</p>
</div></details><h2 id="structs" class="section-header">Structs<a href="#structs" class="anchor">ยง</a></h2><dl class="item-table"><dt><a class="struct" href="struct.ArrayMap.html" title="struct lattices::collections::ArrayMap">Array<wbr>Map</a></dt><dd>An array wrapper representing a fixed-size map.</dd><dt><a class="struct" href="struct.ArraySet.html" title="struct lattices::collections::ArraySet">Array<wbr>Set</a></dt><dd>An array wrapper representing a fixed-size set (modulo duplicate items).</dd><dt><a class="struct" href="struct.EmptyMap.html" title="struct lattices::collections::EmptyMap">Empty<wbr>Map</a></dt><dd>A key-value entry wrapper representing a singleton map.</dd><dt><a class="struct" href="struct.EmptySet.html" title="struct lattices::collections::EmptySet">Empty<wbr>Set</a></dt><dd>A type that will always be an empty set.</dd><dt><a class="struct" href="struct.OptionMap.html" title="struct lattices::collections::OptionMap">Option<wbr>Map</a></dt><dd>A key-value entry wrapper around <code>Option&lt;(K, V)&gt;</code> representing a singleton or empty map.</dd><dt><a class="struct" href="struct.OptionSet.html" title="struct lattices::collections::OptionSet">Option<wbr>Set</a></dt><dd>A wrapper around <code>Option</code>, representing either a singleton or empty set.</dd><dt><a class="struct" href="struct.SingletonMap.html" title="struct lattices::collections::SingletonMap">Singleton<wbr>Map</a></dt><dd>A key-value entry wrapper representing a singleton map.</dd><dt><a class="struct" href="struct.SingletonSet.html" title="struct lattices::collections::SingletonSet">Singleton<wbr>Set</a></dt><dd>A wrapper around an item, representing a singleton set.</dd><dt><a class="struct" href="struct.VecMap.html" title="struct lattices::collections::VecMap">VecMap</a></dt><dd>A <a href="https://doc.rust-lang.org/nightly/alloc/vec/struct.Vec.html" title="struct alloc::vec::Vec"><code>Vec</code></a>-wrapper representing a naively implemented map.</dd><dt><a class="struct" href="struct.VecSet.html" title="struct lattices::collections::VecSet">VecSet</a></dt><dd>A <a href="https://doc.rust-lang.org/nightly/alloc/vec/struct.Vec.html" title="struct alloc::vec::Vec"><code>Vec</code></a>-wrapper representing a naively-implemented set.</dd></dl><h2 id="traits" class="section-header">Traits<a href="#traits" class="anchor">ยง</a></h2><dl class="item-table"><dt><a class="trait" href="trait.MapMapValues.html" title="trait lattices::collections::MapMapValues">MapMap<wbr>Values</a></dt><dd>Trait for transforming the values of a map without changing the overall type of the data structure.</dd></dl></section></div></main></body></html>