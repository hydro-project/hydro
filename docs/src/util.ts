export function getLines(str: string, lineStart: number, lineEnd?: number): string {
    return str.split('\n').slice(lineStart - 1, lineEnd || lineStart).join('\n');
}

export function extractOutput(output: string): string {
    const outputLines = output.replace(/\n$/, '').split('\n');
    outputLines.splice(0, 4);
    if (outputLines[0].startsWith('%%')) {
        const count = outputLines.findIndex(line => 0 === line.length);
        outputLines.splice(0, count + 1, '<graph output>');
    }
    const stdOut = outputLines.join('\n');
    return `#shell-command-next-line
cargo run
<build output>
${stdOut}`;
}

export function extractMermaid(output: string): string {
    const outputLines = output.split('\n');
    outputLines.splice(0, 4);
    const count = outputLines.findIndex(line => 0 === line.length);
    outputLines.length = count;
    return outputLines.join('\n');
}