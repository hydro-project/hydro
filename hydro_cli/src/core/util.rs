use std::time::Duration;

use anyhow::Result;
use futures::Future;

pub async fn async_retry<T, F: Future<Output = Result<T>>>(
    thunk: impl Fn() -> F,
    count: usize,
    delay: Duration,
) -> Result<T> {
    for _ in 1..count {
        let result = thunk().await;
        if result.is_ok() {
            return result;
        } else {
            tokio::time::sleep(delay).await;
        }
    }

    thunk().await
}
