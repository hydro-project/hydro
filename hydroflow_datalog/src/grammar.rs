#[rust_sitter::grammar("datalog")]
#[allow(dead_code)]
pub mod datalog {
    #[rust_sitter::language]
    #[derive(Debug)]
    pub struct Program {
        pub rules: Vec<Declaration>,
    }

    #[derive(Debug, Clone)]
    pub enum Declaration {
        Input(#[rust_sitter::leaf(text = ".input")] (), Ident),
        Output(#[rust_sitter::leaf(text = ".output")] (), Ident),
        Rule(Rule),
    }

    #[derive(Debug, Clone)]
    pub struct Rule {
        pub target: Atom,
        #[rust_sitter::leaf(text = ":-")]
        _from: (),
        #[rust_sitter::repeat(non_empty = true)]
        #[rust_sitter::delimited(
            #[rust_sitter::leaf(text = ",")]
            ()
        )]
        pub sources: Vec<Atom>,
        #[rust_sitter::leaf(text = ".")]
        _dot: Option<()>,
    }

    #[derive(Debug, Clone)]
    pub struct Atom {
        pub name: Ident,
        #[rust_sitter::leaf(text = "(")]
        _l_paren: (),
        #[rust_sitter::delimited(
            #[rust_sitter::leaf(text = ",")]
            ()
        )]
        pub fields: Vec<Ident>,
        #[rust_sitter::leaf(text = ")")]
        _r_paren: (),
    }

    #[derive(Clone, PartialEq, Eq, Hash, Ord, PartialOrd, Debug)]
    pub struct Ident {
        #[rust_sitter::leaf(pattern = r"[a-zA-Z_][a-zA-Z0-9_]*", transform = |s| s.to_string())]
        pub name: String,
    }

    #[rust_sitter::extra]
    struct Whitespace {
        #[rust_sitter::leaf(pattern = r"\s")]
        _whitespace: (),
    }
}
