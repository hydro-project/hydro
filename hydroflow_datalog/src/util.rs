use std::ops::RangeFrom;

pub type Counter = RangeFrom<usize>;
