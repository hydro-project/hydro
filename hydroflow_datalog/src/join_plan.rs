use std::collections::{BTreeMap, HashMap};

use hydroflow_lang::{
    graph::flat_graph::FlatGraph,
    parse::{ArrowConnector, IndexInt, Indexing, Pipeline, PipelineLink},
};
use proc_macro2::Span;
use syn::{self, parse_quote};

use crate::{grammar::datalog::Atom, util::Counter};

/// Captures the tree of joins used to compute contributions from a single rule.
pub enum JoinPlan<'a> {
    /// A single relation without any joins, leaves of the tree.
    Source(&'a Atom),
    /// A join between two subtrees.
    Join(Box<JoinPlan<'a>>, Box<JoinPlan<'a>>),
}

/// Tracks the Hydroflow node that corresponds to a subtree of a join plan.
pub struct IntermediateJoinNode {
    /// The name of the Hydroflow node that this join outputs to.
    pub name: syn::Ident,
    /// If this join node outputs data through a `tee()` operator, this is the index to consume the node with.
    /// (this is only used for cases where we are directly reading a relation)
    pub tee_idx: Option<usize>,
    /// A mapping from variables in the rule to the index of the corresponding element in the flattened tuples this node emits.
    pub variable_mapping: BTreeMap<syn::Ident, usize>,
    /// The type of the flattened tuples this node emits.
    pub tuple_type: syn::Type,
}

enum JoinSide {
    Left,
    Right,
}

impl JoinSide {
    fn index(&self) -> usize {
        match self {
            JoinSide::Left => 0,
            JoinSide::Right => 1,
        }
    }
}

/// Generates a Hydroflow pipeline that transforms some input to a join
/// to emit key-value tuples that can be fed into a join operator.
fn emit_join_input_pipeline(
    // The identifiers of the input node that the key should be populated with.
    identifiers_to_join: &[&syn::Ident],
    // The Hydroflow node that is one side of the join.
    source_expanded: &IntermediateJoinNode,
    // The Hydroflow node for the join operator.
    join_node: &syn::Ident,
    // Whether this node contributes to the left or right side of the join.
    join_side: JoinSide,
    // The Hydroflow graph to emit the pipeline to.
    flat_graph: &mut FlatGraph,
) {
    let hash_keys: Vec<syn::Expr> = identifiers_to_join
        .iter()
        .map(|ident| {
            if let Some(idx) = source_expanded.variable_mapping.get(ident) {
                let idx_ident = syn::Index::from(*idx);
                parse_quote!(v.#idx_ident)
            } else {
                panic!("Could not find key that is being joined on: {:?}", ident);
            }
        })
        .collect();

    let out_index = syn::Index::from(join_side.index());

    let source_name = &source_expanded.name;
    let source_type = &source_expanded.tuple_type;
    flat_graph.add_statement(hydroflow_lang::parse::HfStatement::Pipeline(
        Pipeline::Link(PipelineLink {
            lhs: Box::new(parse_quote!(#source_name)),
            connector: ArrowConnector {
                src: source_expanded.tee_idx.map(|i| Indexing {
                    bracket_token: syn::token::Bracket::default(),
                    index: IndexInt {
                        value: i,
                        span: Span::call_site(),
                    },
                }),
                arrow: parse_quote!(->),
                dst: None,
            },
            rhs: Box::new(parse_quote! {
                map(|v: #source_type| ((#(#hash_keys, )*), v)) -> [#out_index] #join_node
            }),
        }),
    ));
}

/// Generates a Hydroflow pipeline that computes the output to a given [`JoinPlan`].
pub fn expand_join_plan(
    // The plan we are converting to a Hydroflow pipeline.
    plan: &JoinPlan,
    // The Hydroflow graph to emit the pipeline to.
    flat_graph: &mut FlatGraph,
    tee_counter: &mut HashMap<String, Counter>,
    next_join_idx: &mut Counter,
) -> IntermediateJoinNode {
    match plan {
        JoinPlan::Source(target) => {
            let mut variable_mapping = BTreeMap::new();
            let mut row_types: Vec<syn::Type> = vec![];

            // for each variable, a vec of all tuple indices that should equal that variable
            // we only track variables with >= 2 indices, since that is when we need to enforce constraints
            let mut local_constraints = BTreeMap::new();

            for (i, ident) in target.fields.iter().enumerate() {
                row_types.push(parse_quote!(_));
                let variable_ident = syn::Ident::new(&ident.name, Span::call_site());

                // TODO(shadaj): is there something nicer than a clone here?
                match variable_mapping.entry(variable_ident) {
                    std::collections::btree_map::Entry::Vacant(e) => {
                        e.insert(i);
                    }

                    std::collections::btree_map::Entry::Occupied(e) => {
                        let constraint_entry = local_constraints
                            .entry(e.key().clone())
                            .or_insert_with(|| vec![*e.get()]);
                        constraint_entry.push(i);
                    }
                }
            }

            // Because this is a node corresponding to some Datalog relation, we need to tee from it.
            let my_tee_index = tee_counter
                .entry(target.name.name.clone())
                .or_insert_with(|| 0..)
                .next()
                .expect("Out of tee indices");

            let row_type = parse_quote!((#(#row_types, )*));

            if !local_constraints.is_empty() {
                let relation_node = syn::Ident::new(&target.name.name, Span::call_site());
                let relation_idx = syn::Index::from(my_tee_index);

                let filter_node = syn::Ident::new(
                    &format!(
                        "join_{}_filter",
                        next_join_idx.next().expect("Out of join indices")
                    ),
                    Span::call_site(),
                );

                let conditions = local_constraints
                    .values()
                    .map(|indices| {
                        let equal_indices = indices
                            .iter()
                            .map(|i| syn::Index::from(*i))
                            .collect::<Vec<_>>();

                        let first_index = &equal_indices[0];

                        equal_indices
                            .iter()
                            .skip(1)
                            .map(|i| parse_quote!(row.#first_index == row.#i))
                            .reduce(|a: syn::Expr, b| parse_quote!(#a && #b))
                            .unwrap()
                    })
                    .reduce(|a: syn::Expr, b| parse_quote!(#a && #b))
                    .unwrap();

                flat_graph.add_statement(parse_quote! {
                    #filter_node = #relation_node [#relation_idx] -> filter(|&row: &#row_type| #conditions)
                });

                IntermediateJoinNode {
                    name: filter_node,
                    tee_idx: None,
                    variable_mapping,
                    tuple_type: row_type,
                }
            } else {
                IntermediateJoinNode {
                    name: syn::Ident::new(&target.name.name, Span::call_site()),
                    tee_idx: Some(my_tee_index),
                    variable_mapping,
                    tuple_type: row_type,
                }
            }
        }
        JoinPlan::Join(lhs, rhs) => {
            let left_expanded = expand_join_plan(lhs, flat_graph, tee_counter, next_join_idx);
            let right_expanded = expand_join_plan(rhs, flat_graph, tee_counter, next_join_idx);

            let identifiers_to_join = right_expanded
                .variable_mapping
                .keys()
                .filter(|i| left_expanded.variable_mapping.contains_key(i))
                .collect::<Vec<_>>();

            // we start by defining the pipeline from the `join()` operator onwards
            // the main logic here is to flatten the tuples from the left and right sides of the join
            // into a single tuple that is used by downstream joins or the final output
            let mut flattened_tuple_elems: Vec<syn::Expr> = vec![];
            let mut flattened_mapping = BTreeMap::new();

            for (ident, source_idx) in left_expanded
                .variable_mapping
                .keys()
                .map(|l| (l, 0))
                .chain(right_expanded.variable_mapping.keys().map(|l| (l, 1)))
            {
                if !flattened_mapping.contains_key(ident) {
                    let syn_source_index = syn::Index::from(source_idx);
                    let source_expr: syn::Expr = parse_quote!(kv.1.#syn_source_index);
                    let bindings = if source_idx == 0 {
                        &left_expanded.variable_mapping
                    } else {
                        &right_expanded.variable_mapping
                    };

                    let source_col_idx = syn::Index::from(*bindings.get(ident).unwrap());

                    flattened_mapping.insert(ident.clone(), flattened_tuple_elems.len());
                    flattened_tuple_elems.push(parse_quote!(#source_expr.#source_col_idx));
                }
            }

            let key_type = identifiers_to_join
                .iter()
                .map(|_| parse_quote!(_))
                .collect::<Vec<syn::Type>>();

            let left_type = &left_expanded.tuple_type;
            let right_type = &right_expanded.tuple_type;

            let flatten_closure: syn::Expr = parse_quote!(|kv: ((#(#key_type, )*), (#left_type, #right_type))| (#(#flattened_tuple_elems, )*));

            let join_node = syn::Ident::new(
                &format!(
                    "join_{}",
                    next_join_idx.next().expect("Out of join indices")
                ),
                Span::call_site(),
            );
            flat_graph.add_statement(parse_quote!(#join_node = join() -> map(#flatten_closure)));

            emit_join_input_pipeline(
                &identifiers_to_join,
                &left_expanded,
                &join_node,
                JoinSide::Left,
                flat_graph,
            );

            emit_join_input_pipeline(
                &identifiers_to_join,
                &right_expanded,
                &join_node,
                JoinSide::Right,
                flat_graph,
            );

            let output_types: Vec<syn::Type> = flattened_tuple_elems
                .iter()
                .map(|_| parse_quote!(_))
                .collect::<Vec<_>>();

            IntermediateJoinNode {
                name: join_node,
                tee_idx: None,
                variable_mapping: flattened_mapping,
                tuple_type: parse_quote!((#(#output_types, )*)),
            }
        }
    }
}
