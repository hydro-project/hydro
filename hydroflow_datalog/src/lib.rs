use std::collections::{HashMap, HashSet};

use hydroflow_lang::{
    graph::flat_graph::FlatGraph,
    parse::{ArrowConnector, IndexInt, Indexing, Pipeline, PipelineLink},
};
use proc_macro2::{Span, TokenStream};
use quote::{quote, ToTokens};
use syn::parse_quote;

mod grammar;
mod join_plan;
mod util;

use grammar::datalog::*;
use join_plan::*;
use util::Counter;

fn gen_datalog_program(literal: proc_macro2::Literal, root: TokenStream) -> syn::Stmt {
    let str_node: syn::LitStr = parse_quote!(#literal);
    let actual_str = str_node.value();
    let program: Program = grammar::datalog::parse(&actual_str).unwrap();

    let mut inputs = Vec::new();
    let mut outputs = Vec::new();
    let mut rules = Vec::new();

    for stmt in &program.rules {
        match stmt {
            Declaration::Input(_, ident) => inputs.push(ident),
            Declaration::Output(_, ident) => outputs.push(ident),
            Declaration::Rule(rule) => rules.push(rule),
        }
    }

    let mut flat_graph = FlatGraph::default();
    let mut tee_counter = HashMap::new();
    let mut merge_counter = HashMap::new();

    let mut created_rules = HashSet::new();
    for decl in &program.rules {
        let target_ident = match decl {
            Declaration::Input(_, ident) => ident.clone(),
            Declaration::Output(_, ident) => ident.clone(),
            Declaration::Rule(rule) => rule.target.name.clone(),
        };

        if !created_rules.contains(&target_ident) {
            created_rules.insert(target_ident.clone());
            let name = syn::Ident::new(&target_ident.name, Span::call_site());
            flat_graph.add_statement(parse_quote!(#name = merge() -> tee()));
        }
    }

    for target in inputs {
        let target_ident = syn::Ident::new(&target.name, Span::call_site());

        let my_merge_index = merge_counter
            .entry(target.name.clone())
            .or_insert_with(Counter::new)
            .next();

        let my_merge_index_lit =
            syn::LitInt::new(&format!("{}", my_merge_index), Span::call_site());
        let name = syn::Ident::new(&target.name, Span::call_site());

        flat_graph.add_statement(parse_quote! {
            recv_stream(#target_ident) -> [#my_merge_index_lit] #name
        });
    }

    for target in outputs {
        let my_tee_index = tee_counter
            .entry(target.name.clone())
            .or_insert_with(Counter::new)
            .next();

        let out_send_ident = syn::Ident::new(&target.name, Span::call_site());

        let my_tee_index_lit = syn::LitInt::new(&format!("{}", my_tee_index), Span::call_site());
        let target_ident = syn::Ident::new(&target.name, Span::call_site());

        flat_graph.add_statement(parse_quote! {
            #target_ident [#my_tee_index_lit] -> for_each(|v| #out_send_ident.send(v).unwrap())
        });
    }

    let mut next_join_idx = Counter::new();
    for rule in rules {
        generate_rule(
            rule,
            &mut flat_graph,
            &mut tee_counter,
            &mut merge_counter,
            &mut next_join_idx,
        );
    }

    println!("{}", flat_graph.surface_syntax_string());

    let code_tokens = flat_graph
        .into_partitioned_graph()
        .expect("failed to partition")
        .as_code(root);

    syn::parse_quote!({
        #code_tokens
    })
}

fn generate_rule(
    rule: &Rule,
    flat_graph: &mut FlatGraph,
    tee_counter: &mut HashMap<String, Counter>,
    merge_counter: &mut HashMap<String, Counter>,
    next_join_idx: &mut Counter,
) {
    let target = &rule.target.name;
    let target_ident = syn::Ident::new(&target.name, Span::call_site());

    let sources: Vec<Atom> = rule.sources.to_vec();

    // TODO(shadaj): smarter plans
    let plan = sources
        .iter()
        .map(JoinPlan::Source)
        .reduce(|a, b| JoinPlan::Join(Box::new(a), Box::new(b)))
        .unwrap();

    let out_expanded = expand_join_plan(&plan, flat_graph, tee_counter, next_join_idx);

    let output_tuple_elems = rule
        .target
        .fields
        .iter()
        .map(|field| {
            let col = out_expanded
                .variable_mapping
                .get(&syn::Ident::new(&field.name, Span::call_site()))
                .unwrap();
            let source_col_idx = syn::Index::from(*col);

            parse_quote!(row.#source_col_idx)
        })
        .collect::<Vec<syn::Expr>>();

    let flattened_tuple_type = out_expanded.tuple_type;
    let after_join_map: syn::Expr =
        parse_quote!(|row: #flattened_tuple_type| (#(#output_tuple_elems, )*));

    let my_merge_index = merge_counter
        .entry(target.name.clone())
        .or_insert_with(Counter::new)
        .next();

    let my_merge_index_lit = syn::LitInt::new(&format!("{}", my_merge_index), Span::call_site());

    let after_join: Pipeline = parse_quote! {
        map(#after_join_map) -> [#my_merge_index_lit] #target_ident
    };

    let out_name = out_expanded.name;
    flat_graph.add_statement(hydroflow_lang::parse::HfStatement::Pipeline(
        Pipeline::Link(PipelineLink {
            lhs: Box::new(parse_quote!(#out_name)),
            connector: ArrowConnector {
                // if the output comes with a tee index, we must read with that
                // this only happens when we are directly outputting a transformation
                // of a single relation on the RHS
                src: out_expanded.tee_idx.map(|i| Indexing {
                    bracket_token: syn::token::Bracket::default(),
                    index: IndexInt {
                        value: i,
                        span: Span::call_site(),
                    },
                }),
                arrow: parse_quote!(->),
                dst: None,
            },
            rhs: Box::new(after_join),
        }),
    ));
}

#[proc_macro]
pub fn datalog(item: proc_macro::TokenStream) -> proc_macro::TokenStream {
    let item = proc_macro2::TokenStream::from(item);
    let literal: proc_macro2::Literal = syn::parse_quote! {
        #item
    };

    let hydroflow_crate = proc_macro_crate::crate_name("hydroflow")
        .expect("hydroflow should be present in `Cargo.toml`");
    let root = match hydroflow_crate {
        proc_macro_crate::FoundCrate::Itself => quote! { hydroflow },
        proc_macro_crate::FoundCrate::Name(name) => {
            let ident = syn::Ident::new(&name, Span::call_site());
            quote! { #ident }
        }
    };

    proc_macro::TokenStream::from(gen_datalog_program(literal, root).to_token_stream())
}

#[cfg(test)]
mod tests {
    use std::fs::File;
    use std::io::{Read, Write};
    use std::process::Command;

    use quote::ToTokens;
    use syn::parse_quote;
    use tempfile::tempdir;

    use super::gen_datalog_program;

    fn rustfmt_code(code: &str) -> String {
        let dir = tempdir().unwrap();
        let file_path = dir.path().join("temp.rs");
        let mut file = File::create(file_path.clone()).unwrap();

        writeln!(file, "{}", code).unwrap();
        drop(file);

        Command::new("rustfmt")
            .arg(file_path.to_str().unwrap())
            .spawn()
            .unwrap()
            .wait()
            .unwrap();

        let mut file = File::open(file_path).unwrap();
        let mut data = String::new();
        file.read_to_string(&mut data).unwrap();
        drop(file);
        dir.close().unwrap();
        data
    }

    #[test]
    fn minimal_program() {
        let out = &gen_datalog_program(
            parse_quote!(
                r#"
                .input input
                .output out

                out(y, x) :- input(x, y).
                "#
            ),
            quote::quote! { hydroflow },
        );

        let wrapped: syn::Item = parse_quote! {
            fn main() {
                #out
            }
        };

        insta::assert_display_snapshot!(rustfmt_code(&wrapped.to_token_stream().to_string()));
    }

    #[test]
    fn join_with_self() {
        let out = &gen_datalog_program(
            parse_quote!(
                r#"
                .input input
                .output out

                out(x, y) :- input(x, y), input(y, x).
                "#
            ),
            quote::quote! { hydroflow },
        );

        let wrapped: syn::Item = parse_quote! {
            fn main() {
                #out
            }
        };

        insta::assert_display_snapshot!(rustfmt_code(&wrapped.to_token_stream().to_string()));
    }

    #[test]
    fn join_with_other() {
        let out = &gen_datalog_program(
            parse_quote!(
                r#"
                .input in1
                .input in2
                .output out

                out(x, y) :- in1(x, y), in2(y, x).
                "#
            ),
            quote::quote! { hydroflow },
        );

        let wrapped: syn::Item = parse_quote! {
            fn main() {
                #out
            }
        };

        insta::assert_display_snapshot!(rustfmt_code(&wrapped.to_token_stream().to_string()));
    }

    #[test]
    fn multiple_contributors() {
        let out = &gen_datalog_program(
            parse_quote!(
                r#"
                .input in1
                .input in2
                .output out

                out(x, y) :- in1(x, y).
                out(x, y) :- in2(y, x).
                "#
            ),
            quote::quote! { hydroflow },
        );

        let wrapped: syn::Item = parse_quote! {
            fn main() {
                #out
            }
        };

        insta::assert_display_snapshot!(rustfmt_code(&wrapped.to_token_stream().to_string()));
    }

    #[test]
    fn single_column_program() {
        let out = &gen_datalog_program(
            parse_quote!(
                r#"
                .input in1
                .input in2
                .output out

                out(x) :- in1(x), in2(x).
                "#
            ),
            quote::quote! { hydroflow },
        );

        let wrapped: syn::Item = parse_quote! {
            fn main() {
                #out
            }
        };

        insta::assert_display_snapshot!(rustfmt_code(&wrapped.to_token_stream().to_string()));
    }

    #[test]
    fn triple_relation_join() {
        let out = &gen_datalog_program(
            parse_quote!(
                r#"
                .input in1
                .input in2
                .input in3
                .output out

                out(d, c, b, a) :- in1(a, b), in2(b, c), in3(c, d).
                "#
            ),
            quote::quote! { hydroflow },
        );

        let wrapped: syn::Item = parse_quote! {
            fn main() {
                #out
            }
        };

        insta::assert_display_snapshot!(rustfmt_code(&wrapped.to_token_stream().to_string()));
    }
}
