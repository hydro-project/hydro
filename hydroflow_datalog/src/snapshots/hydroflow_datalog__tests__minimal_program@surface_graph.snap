---
source: hydroflow_datalog/src/lib.rs
expression: graph.surface_syntax_string()
---
1v1 = merge ();
2v1 = tee ();
3v1 = merge ();
4v1 = tee ();
5v1 = source_stream (input);
6v1 = for_each (| v | out . send (v) . unwrap ());
7v1 = map (| row : (_ , _ ,) | (row . 1 , row . 0 ,));

(1v1-->2v1);
(3v1-->4v1);
(5v1-->1v1);
(4v1-->6v1);
(7v1-->3v1);
(2v1-->7v1);

